ddtrace/internal/processor/trace.py

```diff
@@ -0,0 +1,127 @@
+import abc
+from collections import defaultdict
+import threading
+from typing import DefaultDict
+from typing import Iterable
+from typing import List
+from typing import Optional
+
+import attr
+import six
+
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.processor import SpanProcessor
+from ddtrace.internal.writer import TraceWriter
+from ddtrace.span import Span
+
+
+log = get_logger(__name__)
+
+
+@attr.s
+class TraceProcessor(six.with_metaclass(abc.ABCMeta)):
+    def __attrs_post_init__(self):
+        # type () -> None
+        """Default post initializer which logs the representation of the
+        TraceProcessor at the ``logging.DEBUG`` level.
+
+        The representation can be modified with the ``repr`` argument to the
+        attrs attribute::
+
+            @attr.s
+            class MyTraceProcessor(TraceProcessor):
+                field_to_include = attr.ib(repr=True)
+                field_to_exclude = attr.ib(repr=False)
+        """
+        log.debug("initialized trace processor %r", self)
+
+    @abc.abstractmethod
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        """Processes a trace.
+
+        None can be returned to prevent the trace from being exported.
+        """
+        pass
+
+
+@attr.s
+class TraceSamplingProcessor(TraceProcessor):
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        sampled_spans = [s for s in trace if s.sampled]
+        if len(sampled_spans) == 0:
+            log.info("dropping trace, %d spans unsampled", len(trace))
+            return None
+
+        log.info("trace %d sampled (%d/%d spans sampled)", trace[0].trace_id, len(sampled_spans), len(trace))
+        return trace
+
+
+@attr.s
+class SpansToTraceProcessor(SpanProcessor):
```

Comment:
How do you feel about `SpanAggregator` instead of `SpansToTraceProcessor`? The `Trace` bit actually refers to a different `Trace` object than the one in `ddtrace.tracer` and could be a bit confusing?

Reply:
Yes I like it! I don't like the name trace because it doesn't have a strong definition in general.